### PR TITLE
Added basic logging with rich handler

### LINKED ISSUE #35: [Get Familiar]: Improve logging
Some ideas for logs: 

1.  Explore the code based on the logs you see.
2. Add log levels to them as part of your first contribution! 
3. If tidy readable outputs interest you, try using colorama to indicate log levels
4. You can also propose "modes" of logs: dev/user/minimal/silent

**Note:** this issue is marked with the `hacktoberfest` label. Follow `#hacktoberfest-discussions` [on Discord](https://discord.gg/ANUuzsrpVV) for further details .

### PR DESCRIPTION
**Pull request by pixalquarks**
Closes  #35 

### Added basic logging with rich handler

# Features
- implements global level logs, with all main logger implemented at a single file.
- support for print type argument passing. (multiple data types and sep, end variable).

## Log Previews

Screenshot (41)
Screenshot (40)

### CODE REVIEW

File: src/logger.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import logging
+from rich.logging import RichHandler
+
+FORMAT = "%(message)s"
+
+class Logger:
+    def __init__(self):
+        self.__FORMAT = "%(message)s"
+        logging.basicConfig(
+                level="NOTSET", format=FORMAT, datefmt="[%X]", handlers=[RichHandler(rich_tracebacks=True)]
```
Comment (ayushjain2809) [inline src/logger.py:10]: `rich_tracebacks` can cause a performance issue. Allow users to disable it via configuration

File: src/logger.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                )
+        self.__logger = logging.getLogger("rich")
```
Comment (ayushjain2809) [inline src/logger.py:12]: why use "rich" as logger name? It should point to where in application it originated. We can use "main" or "logger" (__name__)

File: src/logger.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        
+    def stringfy(func):
```
Comment (ayushjain2809) [inline src/logger.py:14]: typo stringfy -> stringify

File: src/logger.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            for v in msg:
+                if not isinstance(v, str):
+                    v = str(v)
+                nmsg.append(v)
+            func(self, *nmsg, sep=sep, end=end)
+        return inner
+                
+    @stringfy
+    def debug(self, *msg, sep=' ', end='\n'):
+        self.__logger.debug(sep.join(msg))
```
Comment (ayushjain2809) [inline src/logger.py:26]: can you add a comment describing what it does? why @stringify is needed? Why not use self.__logger.debug(f"{msg}") instead?

File: src/processors/CropOnMarkers.py [hunk 5]
```diff
@@ -30,7 +31,7 @@ def __init__(self, marker_ops, cwd):
         self.marker_rescale_steps = marker_ops.get("marker_rescale_steps", 10)
         self.apply_erode_subtract = marker_ops.get("apply_erode_subtract", 1)
         if not os.path.exists(self.marker_path):
-            print(
+            logger.error(
```
Comment (ayushjain2809) [inline src/processors/CropOnMarkers.py:34]: now how about removing the "Error: " prefix?

File: src/processors/CropOnMarkers.py [hunk 6]
```diff
@@ -95,14 +96,14 @@ def getBestMatch(self, image_eroded_sub):
                 best_scale, all_max_t = s, max_t
 
         if all_max_t < self.min_matching_threshold:
-            print(
+            logger.warning(
```
Comment (ayushjain2809) [inline src/processors/CropOnMarkers.py:99]: keep '\t', remove "Warning:"

Comment (pixalquarks) [reply src/processors/CropOnMarkers.py:99]: fixed

File: src/processors/CropOnMarkers.py (hunk 5 above)
Comment (pixalquarks) [reply src/processors/CropOnMarkers.py:34]: fixed

File: src/logger.py (hunk 1 above)
Comment (pixalquarks) [reply src/logger.py:10]: done

File: src/logger.py (hunk 2 above)
Comment (pixalquarks) [reply src/logger.py:12]: done. Btw should I add module level logging, or a single global logger is good enough?

File: src/logger.py (hunk 3 above)
Comment (pixalquarks) [reply src/logger.py:14]: done

File: src/logger.py (hunk 4 above)
Comment (pixalquarks) [reply src/logger.py:26]: It was done to provide a print like syntax for the logging functions, we can remove it, if it's not needed.

File: src/logger.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            for v in msg:
+                if not isinstance(v, str):
+                    v = str(v)
+                nmsg.append(v)
+            func(self, *nmsg, sep=sep, end=end)
+        return inner
+                
+    @stringify
+    def debug(self, *msg, sep=' ', end='\n'):
+        self.log.debug(sep.join(msg), stacklevel=3)  # set stack level to 3 so that the caller of this function is logged, not this function itself. stack-frame - self.log.debug - stringify:28 - caller
```
Comment (Udayraj123) [inline src/logger.py:33]: move comment above the line. Line width should be within limits (can create another issue for stricter python standards)

Comment (Udayraj123) [reply src/logger.py:33]: also, sep.join and stacklevel=3 is getting repeated in all of them. Move it into a util function - 
```
@stringify
def logutil(self, methodType, *msg, sep, end): 
      return self.log[methodType](sep.join(msg), stacklevel=3)

# no stringify needed here
def debug(*args):
    return logutil('debug', *args)
```

**Review by Udayraj123 — DISMISSED** [review]

**Review by Udayraj123 — CHANGES_REQUESTED** [review]
comments

File: src/logger.py [hunk 8]
```diff
@@ -15,37 +15,40 @@ def __init__(self, name, level: Union[int, str] = logging.NOTSET,  message_forma
...
+    @stringify 
+    def logutil(self, method_type, *msg, sep):
+        func = getattr(self.log, method_type, __default = None)
+        if not func:
+            return
+        func(sep.join(msg), stacklevel=4)
+    
     def debug(self, *msg, sep=' ', end='\n'):
-        self.log.debug(sep.join(msg), stacklevel=3)  # set stack level to 3 so that the caller of this function is logged, not this function itself. stack-frame - self.log.debug - stringify:28 - caller
+        self.logutil('debug', *msg, sep=sep)
```
Comment (Udayraj123) [inline src/logger.py:40]: a good practice to add return here

Comment (Udayraj123) [reply src/logger.py:40]: Also can we set default text colors inside this util now?

File: src/logger.py (hunk 2 above)
Comment (Udayraj123) [reply src/logger.py]: Yeah global logger is all fine. We just require ability to reduce log to INFO level if needed

File: src/logger.py (hunk 8 above)
Comment (Udayraj123) [reply src/logger.py]: Cool

File: src/logger.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import logging
+from typing import Iterable, Union, List, Any, Callable, Tuple
+from rich.logging import RichHandler
+from .config import CONFIG_DEFAULTS as config
+
+FORMAT = "%(message)s"
+
+logging.basicConfig(
+                level=logging.NOTSET, format="%(message)s", datefmt="[%X]", handlers=[RichHandler(rich_tracebacks=config.outputs.rich_tracebacks)]
```
Comment (Udayraj123) [inline src/logger.py:9]: hmm, even better now

File: src/logger.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        return inner
+    
+    # set stack level to 3 so that the caller of this function is logged, not this function itself.
+    # stack-frame - self.log.debug - logutil - stringify - log method - caller
+    @stringify 
+    def logutil(self, method_type: str, *msg: object, sep=' ') -> None:
+        func = getattr(self.log, method_type, None)
+        if not func:
+            raise AttributeError(f"Logger has no method {method_type}")
+        return func(sep.join(msg), stacklevel=4)
```
Comment (Udayraj123) [inline src/logger.py:36]: add spaces around '=' at all places e.g. `stacklevel = 4`

**Review by Udayraj123 — APPROVED** [review]
Approved with comment.

File: src/logger.py (hunk 10 above)
Comment (pixalquarks) [reply src/logger.py:36]: Spaces are not recommended around keyword arguments according to PEP.

Comment (Udayraj123) [reply src/logger.py:36]: odd, but true. cool

**Comment by Udayraj123** [comment]
@pixalquarks can you please take latest pull and run `python3 main.py -i samples/` and share its output screenshots below?

**Comment by pixalquarks** [comment]
Screenshot (47)
Screenshot (48)
Screenshot (49)
Screenshot (50)
Screenshot (51)
Screenshot (52)
Screenshot (53)
Screenshot (54)
Screenshot (55)
Screenshot (56)
Screenshot (57)
Screenshot (58)
Screenshot (59)
Screenshot (60)
Screenshot (61)
Screenshot (62)
Screenshot (63)
Screenshot (64)
Screenshot (65)

**Comment by Udayraj123** [comment]
Can you make changes for "Validating Template" and the line below? Take the latest pull

**Comment by pixalquarks** [comment]
Screenshot (66)

**Comment by Gauravmauryaios** [comment]
var restify = require('restify')

// config vars
var FB_LOGIN_PATH    = '/api/facebook_login'
var FB_CALLBACK_PATH = '/api/facebook_callback'
var FB_APPID = '<<YOUR APPID HERE>>'
var FB_APPSECRET = '<<YOUR APPSECRET HERE>>'
var SERVER_PREFIX = 'http://localhost:3000'

// set up server
var server = restify.createServer()
server.use(restify.queryParser());

// set up passport-facebook
var passport         = require('passport')
  , FacebookStrategy = require('passport-facebook').Strategy;

var fb_login_handler    = passport.authenticate('facebook')
var fb_callback_handler = passport.authenticate('facebook') 
var fb_callback_handler2 = function(req, res) {
    console.log('we b logged in!')
    console.dir(req.user)
}

server.get(FB_LOGIN_PATH,    fb_login_handler)
server.get(FB_CALLBACK_PATH, fb_callback_handler, fb_callback_handler2)

passport.use(new FacebookStrategy({
    clientID:     FB_APPID,
    clientSecret: FB_APPSECRET,
    callbackURL:  SERVER_PREFIX + FB_CALLBACK_PATH
  },
  function(accessToken, refreshToken, profile, done) {
      console.log('accessToken='+accessToken+' facebookId='+profile.id)
      return done(null, profile)
  })
)

// Start the app by listening on <port>
var port =  process.env.PORT || 3000
server.listen(port)
console.log('App started on port ' + port)``

**Comment by Udayraj123** [comment]
@Gauravmauryaios accidental comment? can delete from here.

**Pull request merged**